x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/index.tsx

```diff
@@ -165,7 +166,20 @@ const StepDefineRuleComponent: FC<StepDefineRuleProps> = ({
   // reset form when rule type changes
   useEffect(() => {
     reset({ resetValues: false });
-  }, [reset, ruleType]);
+    if (getFields().queryBar != null) {
+      const queryBar = getFields().queryBar.value as DefineStepRule['queryBar'];
+      if (queryBar.edited !== true && !isUpdateView) {
```

Comment:
FYI this `edited` field is mostly redundant with the built in `isPristine` property on each field, except that `isPristine` becomes `false`  when the value changes below, while `edited` does not.

Perhaps we could use `field.isPristine || field.value.query.query === '*:*'` to accomplish the same logic, and do away with the `edited` property?

Reply:
In talking with @FrankHassanabad I wanted to add this to our team discussion. It's becoming increasingly more difficult to make any changes without either 1) breaking another rule type's logic or 2) adding extra validations throughout. It's feeling like a bit of wack-a-mole. I almost want to just silo off each rule type. One example I ran into with preview was that each rule held the state of other rule types (if let's say I visited Threshold and set some state and then switched to custom query) and had to add logic to ensure that it was picking up and ignoring the right parts.

Reply:
@yctercero Would the schemas [here](https://github.com/elastic/kibana/blob/e7cbd53924e97aa58096650b5064d87a475004da/x-pack/plugins/security_solution/common/detection_engine/schemas/request/rule_schemas.ts#L343) help to silo off the rule types in the frontend code? I created them with a similar goal in mind for the backend of making it clear which fields are valid for each rule type, as well as defining which fields are shared across all rule types. They're usable in the frontend as well but I'm not sure if they fully implement the siloing you have in mind.

Reply:
@marshallmain yea, I think they would. It might add more touchpoints in the front, but that may be worth the security of not having to worry that an edit to one rule will break another.

Reply:
@ryland, The `isPristine` field is something I want to use, but there are more odd side effects of how we use the forms where we do a `reset({ resetValues: false });` which will reset the isPristine back to true. Even in that block of code if I were to do something like:

```ts
  // reset form when rule type changes
  useEffect(() => {
    reset({ resetValues: false });
    if (getFields().queryBar != null) {
      const queryBar = getFields().queryBar.value as DefineStepRule['queryBar'];
      if (queryBar.query.query != null) {
        if (
          isThreatMatchRule(ruleType) &&
          (getFields().queryBar.isPristine || queryBar.query.query === '')
        ) {
          setFieldValue('queryBar', {
            ...stepDefineDefaultValue.queryBar,
            query: { ...stepDefineDefaultValue.queryBar.query, query: '*:*' },
          });
        } else if (getFields().queryBar.isPristine || queryBar.query.query === '*:*') {
          setFieldValue('queryBar', stepDefineDefaultValue.queryBar);
        } else {
          setFieldValue('queryBar', queryBar);
        }
      }
    }
  }, [reset, ruleType, setFieldValue, getFields, isUpdateView]);
```

It still won't work all the way because `isPristine` will reset back to true and stay true based on two things which is:

* The reset will turn it back to `true`
* Even if this block of code in the else tries to force pristine back to false:
```ts
else {
  setFieldValue('queryBar', queryBar);
}
```
The forms code if it sees the queryBar value is the same as before it will not trigger pristine _back_ to false. So pristine becomes not useful here.

However, to reduce state at a trade off of the forms having funny quirks if a user decides to use `*:*` in a different rule if they're really flipping around between rules a lot before going to the next step would be this:

```ts
  // reset form when rule type changes
  useEffect(() => {
    reset({ resetValues: false });
    const queryBar = getFields().queryBar.value as DefineStepRule['queryBar'];
    if (queryBar.query.query != null) {
      if (isThreatMatchRule(ruleType) && queryBar.query.query === '') {
        setFieldValue('queryBar', {
          ...stepDefineDefaultValue.queryBar,
          query: { ...stepDefineDefaultValue.queryBar.query, query: '*:*' },
        });
      } else if (queryBar.query.query === '*:*') {
        setFieldValue('queryBar', stepDefineDefaultValue.queryBar);
      }
    }
  }, [reset, ruleType, setFieldValue, getFields, isUpdateView]);
```

That would just hardcode the two conditions of either detecting an empty string or detecting a `*:*` but if a user decides to use `*:*` and then flip between rules such as threshold and custom query they are going to get the surprise that the `*:*` turns _back_ into empty string. But the validation would turn on to warn them they have an empty string they need to deal with.

If that is deemed the "the lesser of evils" with what we already have in the forms code I can commit the code above.

Reply:
Actually after messing around with it @rylnd I was able to accomplish everything the edited flag could using a combination of the `isPristine` along with `reset` and then the `setValue` that is looking pretty good. I think this is what I am going to go with:

```ts
  // reset form when rule type changes
  useEffect(() => {
    reset({ resetValues: false });
    const { queryBar } = getFields();
    if (isThreatMatchRule(ruleType) && queryBar.isPristine) {
      queryBar.reset({
        defaultValue: {
          ...stepDefineDefaultValue.queryBar,
          query: { ...stepDefineDefaultValue.queryBar.query, query: '*:*' },
        },
      });
    } else if (queryBar.isPristine) {
      queryBar.reset({
        defaultValue: {
          ...stepDefineDefaultValue.queryBar,
          query: { ...stepDefineDefaultValue.queryBar.query, query: '' },
        },
      });
    } else {
      // This ensures the isPristine is still set to false after a user has modified it
      queryBar.setValue((prev: DefineStepRule['queryBar']) => ({ ...prev }));
    }
  }, [reset, ruleType, setFieldValue, getFields, isUpdateView]);
```